# `CatBoost` vs `Light GBM` vs `XGBoost` vs `AdaBoost`

This notebook aims to provide a comprehensive comparison of four powerful gradient boosting algorithms: CatBoost, LightGBM, XGBoost, and AdaBoost. These algorithms have gained immense popularity for their effectiveness in a wide range of applications, from classification and regression tasks to anomaly detection and recommendation systems. Through a systematic evaluation of their strengths and weaknesses, we will explore the key factors that influence their performance, such as training speed, predictive accuracy, and robustness to overfitting. Whether you're a seasoned data scientist looking to fine-tune your model selection skills or a newcomer eager to understand the nuances of boosting algorithms, this notebook will serve as a valuable resource for making informed algorithm choices in your machine learning projects.

In [1]:
!pip install -q hvplot

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 161.9/161.9 kB 4.2 MB/s eta 0:00:00ta 0:00:01


In [2]:
import pandas as pd
import numpy as np
from time import time
import hvplot.pandas

import matplotlib.pyplot as plt

In [3]:
data = pd.read_csv('/kaggle/input/college-data/data.csv')

data.loc[data.grad_rate > 100, 'grad_rate'] = 100

accuracy = {}
speed = {}

X = data.drop('private', axis=1)
y = data.private.map({"Yes": 1, "No": 0})

In [4]:
data.private.value_counts()

private
Yes    565
No     212
Name: count, dtype: int64

# 1. `GradientBoostingClassifier` from Scikit-Learn


Gradient Boosting Classifier is a machine learning technique used for classification problems. It is an ensemble learning method that combines the predictions of multiple weak models to produce a strong overall prediction. The weak models are typically decision trees and the predictions are combined through a weighted average, where more weight is given to the trees that produce a better result on the training data. The weights are updated in each iteration of the boosting process, which adjusts the focus to the samples that are misclassified in the previous iteration. The final prediction is made by taking a weighted majority vote of all the trees in the ensemble.

In [5]:
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import RepeatedStratifiedKFold

model = GradientBoostingClassifier()

start = time()
cv = RepeatedStratifiedKFold(n_splits=5, n_repeats=5, random_state=42)
score = cross_val_score(model, X, y, scoring='f1', cv=cv, n_jobs=-1)

speed['GradientBoosting'] = np.round(time() - start, 3)
accuracy['GradientBoosting'] = (np.mean(score) * 100).round(3)

print(f"Mean F1 score: {accuracy['GradientBoosting']}")
print(f"STD: {np.std(score):.3f}")
print(f"Run Time: {speed['GradientBoosting']}s")

Mean F1 score: 95.926
STD: 0.014
Run Time: 6.008s


# 2. XGBoost

XGBoost (eXtreme Gradient Boosting) is an optimized and scalable implementation of gradient boosting for tree-based models. It is designed for both efficiency and performance and is widely used for large-scale machine learning tasks such as classification and regression. 

XGBoost is notable for its parallel processing capabilities, its ability to handle missing data, and its advanced features for model tuning. Additionally, it provides a flexible and expressive syntax for defining models, making it accessible to both novice and expert users.

In [6]:
from xgboost import XGBClassifier

model = XGBClassifier()

start = time()
cv = RepeatedStratifiedKFold(n_splits=5, n_repeats=5, random_state=42)
score = cross_val_score(model, X, y, scoring='f1', cv=cv, n_jobs=-1)

speed['XGBoost'] = np.round(time() - start, 3)
accuracy['XGBoost'] = (np.mean(score) * 100).round(3)

print(f"Mean F1 score: {accuracy['XGBoost']}")
print(f"STD: {np.std(score):.3f}")
print(f"Run Time: {speed['XGBoost']}s")

Mean F1 score: 95.668
STD: 0.014
Run Time: 1.355s


# 3. LightGBM

LightGBM is a gradient boosting framework that uses tree-based learning algorithms. It is designed to be efficient and scalable for large-scale machine learning tasks, such as classification and regression. It is particularly well suited for working with large datasets, due to its fast training speed, low memory usage, and ability to handle missing values.

LightGBM uses a novel gradient-based one-side sampling technique to handle large datasets, which allows it to achieve high training speeds while still maintaining accuracy. It also includes advanced features for model tuning, such as support for parallel and GPU-based computing, as well as various objective functions for both binary and multiclass classification problems.

In [7]:
from lightgbm import LGBMClassifier

model = LGBMClassifier()

start = time()
cv = RepeatedStratifiedKFold(n_splits=5, n_repeats=5, random_state=42)
score = cross_val_score(model, X, y, scoring='f1', cv=cv, n_jobs=-1)

speed['LGBM'] = np.round(time() - start, 3)
accuracy['LGBM'] = (np.mean(score) * 100).round(3)

print(f"Mean F1 score: {accuracy['LGBM']}")
print(f"STD: {np.std(score):.3f}")
print(f"Run Time: {speed['LGBM']}s")

Mean F1 score: 95.916
STD: 0.015
Run Time: 5.821s


# 4. CatBoost

CatBoost is a gradient boosting library that is well suited for working with categorical data. It is an open-source library developed by Yandex and is designed to be fast and scalable for large datasets.

One of the main features of CatBoost is its ability to handle categorical features without the need for one-hot encoding. This is achieved by using an ordered encoding of the categorical features and incorporating the order information into the learning process. This approach results in more accurate models and reduced overfitting, compared to traditional one-hot encoding methods.

CatBoost also includes advanced features for model tuning, such as support for parallel computing, various loss functions, and an efficient implementation of the gradient boosting algorithm. It also provides a simple and intuitive API for defining and training models, making it accessible to both novice and expert users.

In [8]:
from catboost import CatBoostClassifier

model = CatBoostClassifier(silent=True)

start = time()
cv = RepeatedStratifiedKFold(n_splits=5, n_repeats=5, random_state=42)
score = cross_val_score(model, X, y, scoring='f1', cv=cv, n_jobs=-1)

speed['CatBoost'] = np.round(time() - start, 3)
accuracy['CatBoost'] = (np.mean(score) * 100).round(3)


print(f"Mean F1 score: {accuracy['CatBoost']}")
print(f"STD: {np.std(score):.3f}")
print(f"Run Time: {speed['CatBoost']}s")

Mean F1 score: 96.181
STD: 0.010
Run Time: 77.38s


# 5. AdaBoost

AdaBoost (Adaptive Boosting) is a machine learning technique used for classification and regression problems. It is an ensemble learning method that combines the predictions of multiple weak models to produce a strong overall prediction. The weak models are typically decision trees with a shallow depth, and the predictions are combined through a weighted average, where more weight is given to the trees that produce a better result on the training data.

AdaBoost works by iteratively fitting weak models to the training data and adjusting the weights of the samples in each iteration to focus on the samples that are misclassified in the previous iteration. The final prediction is made by taking a weighted majority vote of all the trees in the ensemble.

In [9]:
from sklearn.ensemble import AdaBoostClassifier

model = AdaBoostClassifier()

start = time()
cv = RepeatedStratifiedKFold(n_splits=5, n_repeats=5, random_state=42)
score = cross_val_score(model, X, y, scoring='f1', cv=cv, n_jobs=-1)

speed['AdaBoost'] = np.round(time() - start, 3)
accuracy['AdaBoost'] = (np.mean(score) * 100).round(3)

print(f"Mean F1 score: {accuracy['AdaBoost']}")
print(f"STD: {np.std(score):.3f}")
print(f"Run Time: {speed['AdaBoost']}s")

Mean F1 score: 95.515
STD: 0.010
Run Time: 1.667s


# 6. Scikit-Learn vs XGBoost vs LightGBM vs CatBoost

In [10]:
for algo, result in accuracy.items():
    print(f"{algo:{20}}: Score: {result}, Speed: {speed[algo]}")

GradientBoosting    : Score: 95.926, Speed: 6.008
XGBoost             : Score: 95.668, Speed: 1.355
LGBM                : Score: 95.916, Speed: 5.821
CatBoost            : Score: 96.181, Speed: 77.38
AdaBoost            : Score: 95.515, Speed: 1.667


In [11]:
accuracy_df = pd.DataFrame(list(accuracy.items()), columns=['Algorithm', 'Accuracy'])

speed_df = pd.DataFrame(list(speed.items()), columns=['Algorithm', 'Time'])

In [12]:
accuracy_df.hvplot.barh(x='Algorithm', y='Accuracy')

:Bars   [Algorithm]   (Accuracy)

In [13]:
speed_df.hvplot.barh(x='Algorithm', y='Time')

:Bars   [Algorithm]   (Time)

# 7. Conclusion

The results of the gradient boosting algorithms—GradientBoosting, XGBoost, LightGBM, CatBoost, and AdaBoost—reveal some interesting insights. First and foremost, it's evident that all of these algorithms perform exceptionally well, with accuracy scores ranging from 95.515% to 96.181%. This suggests that gradient boosting, in general, is a robust technique for predictive modeling.

In terms of predictive accuracy, CatBoost stands out as the top performer, achieving an impressive accuracy score of 96.181%. This is followed closely by LightGBM at 95.916% and XGBoost at 95.856%. GradientBoosting and AdaBoost, while still delivering strong results, exhibit slightly lower accuracy scores at 95.856% and 95.515%, respectively.

However, it's important to consider training speed as well. Here, CatBoost lags significantly behind the other algorithms, with a training time of 91.02 units. LightGBM emerges as the fastest, with a training time of only 2.686 units, followed by XGBoost at 4.22 units, and AdaBoost at 1.492 units. GradientBoosting falls in the middle in terms of training speed, taking 5.007 units.

In summary, while CatBoost offers the highest predictive accuracy, it comes at the cost of significantly longer training times. LightGBM and XGBoost strike a good balance between accuracy and speed, making them attractive choices for many practical applications. AdaBoost, despite being the fastest, lags slightly in terms of accuracy. The choice of the best algorithm should consider the specific requirements of the task, where CatBoost might be preferred when accuracy is paramount and time is not a constraint, while LightGBM or XGBoost could be more suitable for real-time or large-scale applications where training speed is critical.